pkg/controllers/user/istio/appHandler.go

```diff
@@ -114,16 +115,46 @@ func (ah *appHandler) sync(key string, obj *v3.App) (runtime.Object, error) {
 		return obj, nil
 	}
 
-	if obj.DeletionTimestamp != nil {
-		err := isIstioMetricExpressionWithdraw(ah.clusterName, ah.istioClusterGraphClient, ah.istioMonitorMetricClient, obj)
+	if obj == nil || obj.DeletionTimestamp != nil {
+		err := ah.updateClusterIstioCondition(false)
+		if err != nil {
+			return obj, err
+		}
+		return obj, isIstioMetricExpressionWithdraw(ah.clusterName, ah.istioClusterGraphClient, ah.istioMonitorMetricClient, obj)
+	}
+
+	if err := ah.updateClusterIstioCondition(true); err != nil {
 		return obj, err
 	}
 
 	if v3.AppConditionInstalled.IsFalse(obj) {
 		return obj, errors.Errorf("waiting for the app %s to be installed", obj.Name)
 	}
 
-	err := isIstioMetricExpressionDeployed(ah.clusterName, ah.istioClusterGraphClient, ah.istioMonitorMetricClient, obj)
+	return obj, isIstioMetricExpressionDeployed(ah.clusterName, ah.istioClusterGraphClient, ah.istioMonitorMetricClient, obj)
+}
 
-	return obj, err
+func (ah *appHandler) updateClusterIstioCondition(enabled bool) error {
+	var err error
+	retryCount := 5
+	for i := 0; i < retryCount; i++ {
```

Comment:
Why do you need a retry loop? we typically reserve that for the few scenarios where we have to give up and the object would not get re-enqueued.

But the caller to this just returns the error out of the handler, which means the framework would cause the handler to get reran

Reply:
From @orangedeng 

There is a small chance you update cluster object fails. This is the way to ensure to update cluster object successfully.

Just don’t want to block other app handler, especially the helm handler, when updating cluster object fails.

Reply:
I think it's safer to fall back to the framework for the retries. I feel like having retry count 20 can block other critical cluster controllers for a while

Reply:
Removed the retry loop